<!--
  2 failure:  :disappointed_rel..., :microscope: Ther...
  0 warning: 
  2 messages
  
-->

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  :disappointed_relieved: `npm test` failed. See below for details.
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  :microscope: There are library changes, but no changes to the unit tests. That's OK as long as you're refactoring existing code, but will require an admin to merge this PR. Please see <a href="https://github.com/miniman42/titanium_mobile/blob/7_4_X_TF_NOTIFICATIONS_FIXES/README.md#unit-tests">README.md#unit-tests</a> for docs on unit testing.
  </td>
    </tr>
  </tbody>
</table>



<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>

  ```
> titanium-mobile@7.4.1 test /Users/build/jenkins/workspace/dk_titanium_mobile_PR-10328-J2UQ2DNZYF5JA7SJI6HVWN3RRLATYACVDXTPDU4WRQN7SDUC5WVQ
> grunt

Running "appcJs:src:lintOnly" (appcJs) task

Running "eslint:src" (eslint) task

/Users/build/jenkins/workspace/dk_titanium_mobile_PR-10328-J2UQ2DNZYF5JA7SJI6HVWN3RRLATYACVDXTPDU4WRQN7SDUC5WVQ/iphone/cli/commands/_build.js
  202:114  warning  'arr' is defined but never used  no-unused-vars

✖ 1 problem (0 errors, 1 warning)


Running "checkFormat:ios" (checkFormat) task
Fatal error: Error: Formatting incorrect on "iphone/Classes/TiApp.m", proposed changes: <?xml version='1.0'?>
<replacements xml:space='preserve' incomplete_format='false'>
<replacement offset='20927' length='5'>&#10;  </replacement>
<replacement offset='21039' length='5'>&#10;  </replacement>
<replacement offset='21154' length='5'>&#10;  </replacement>
<replacement offset='21182' length='7'>&#10;    </replacement>
<replacement offset='21326' length='5'>&#10;  </replacement>
<replacement offset='21339' length='7'>&#10;    </replacement>
<replacement offset='21466' length='7'>&#10;    </replacement>
<replacement offset='21529' length='5'>&#10;  </replacement>
<replacement offset='21963' length='0'> </replacement>
<replacement offset='22082' length='0'> </replacement>
<replacement offset='22419' length='7'>&#10;    </replacement>
<replacement offset='22905' length='0'> </replacement>
<replacement offset='22906' length='8'>&#10;      </replacement>
<replacement offset='22975' length='0'> </replacement>
<replacement offset='57802' length='5'>&#10;  </replacement>
<replacement offset='57814' length='0'> </replacement>
<replacement offset='57815' length='1'></replacement>
<replacement offset='57906' length='5'>&#10;  </replacement>
<replacement offset='57973' length='9'>&#10;    </replacement>
<replacement offset='58020' length='9'>&#10;    </replacement>
<replacement offset='58084' length='0'> </replacement>
<replacement offset='58123' length='9'>&#10;    </replacement>
<replacement offset='58214' length='9'>&#10;    </replacement>
<replacement offset='58262' length='13'>&#10;      </replacement>
<replacement offset='58338' length='9'>&#10;    </replacement>
<replacement offset='58348' length='9'>&#10;    </replacement>
<replacement offset='58431' length='1'></replacement>
<replacement offset='58433' length='0'> </replacement>
<replacement offset='58445' length='10'>&#10;&#10;    </replacement>
<replacement offset='58474' length='0'> </replacement>
<replacement offset='58475' length='1'></replacement>
<replacement offset='58557' length='9'>&#10;    </replacement>
<replacement offset='58604' length='9'>&#10;    </replacement>
<replacement offset='58659' length='30'>&#10;                         </replacement>
<replacement offset='58739' length='9'>&#10;    </replacement>
<replacement offset='58870' length='9'>&#10;    </replacement>
<replacement offset='58886' length='0'> </replacement>
<replacement offset='58887' length='2'></replacement>
<replacement offset='59044' length='1'></replacement>
<replacement offset='59046' length='9'>&#10;    </replacement>
<replacement offset='59130' length='5'>&#10;  </replacement>
<replacement offset='59143' length='9'>&#10;    </replacement>
<replacement offset='59286' length='5'>&#10;  </replacement>
</replacements>

npm ERR! Test failed.  See above for more details.

```
  </td>
    </tr>
  
<tr>
      <td>:book:</td>
      <td>

  :confetti_ball: Welcome to the Titanium SDK community, miniman42! Thank you so much for your PR, you're helping us make Titanium better. :gift:
  </td>
    </tr>
  </tbody>
</table>


<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

--- a/iphone/Classes/TiApp.m
+++ b/iphone/Classes/TiApp.m
@@ -16,6 +16,7 @@
 #import "TiExceptionHandler.h"
 #import "Webcolor.h"
 #import <AVFoundation/AVFoundation.h>
+#import <CoreLocation/CoreLocation.h>
 #import <CoreSpotlight/CoreSpotlight.h>
 #import <QuartzCore/QuartzCore.h>
 #import <libkern/OSAtomic.h>
@@ -1576,11 +1577,34 @@ TI_INLINE void waitForMemoryPanicCleared(); //WARNING: This must never be run on
   [event setObject:NULL_IF_NIL([[[notification request] content] title]) forKey:@"alertTitle"];
   [event setObject:NULL_IF_NIL([[[notification request] content] subtitle]) forKey:@"alertSubtitle"];
   [event setObject:NULL_IF_NIL([[[notification request] content] launchImageName]) forKey:@"alertLaunchImage"];
-  [event setObject:NULL_IF_NIL([[[notification request] content] sound]) forKey:@"sound"];
   [event setObject:NULL_IF_NIL([[[notification request] content] badge]) forKey:@"badge"];
   [event setObject:NULL_IF_NIL([[[notification request] content] userInfo]) forKey:@"userInfo"];
   [event setObject:NULL_IF_NIL([[[notification request] content] categoryIdentifier]) forKey:@"category"];
   [event setObject:NULL_IF_NIL([[notification request] identifier]) forKey:@"identifier"];
+
+  // iOS 10+ does have "soundName" but "sound" which is a native object. But if we find
+  // a sound in the APS dictionary, we can provide that one for parity
+  if (notification.request.content.userInfo[@"aps"] && notification.request.content.userInfo[@"aps"][@"sound"]) {
+    [event setObject:notification.request.content.userInfo[@"aps"][@"sound"] forKey:@"sound"];
+  }
+
+  // Inject the trigger (time- or location-based) into the payload
+  UNNotificationTrigger *trigger = notification.request.trigger;
+
+  if (trigger != nil) {
+    if ([trigger isKindOfClass:[UNCalendarNotificationTrigger class]]) {
+      [event setObject:NULL_IF_NIL([(UNCalendarNotificationTrigger *)trigger nextTriggerDate]) forKey:@"date"];
+    } else if ([trigger isKindOfClass:[UNLocationNotificationTrigger class]]) {
+      CLCircularRegion *region = (CLCircularRegion *)[(UNLocationNotificationTrigger *)trigger region];
+      NSDictionary *dict = @{
+        @"latitude" : NUMDOUBLE(region.center.latitude),
+        @"longitude" : NUMDOUBLE(region.center.longitude),
+        @"radius" : NUMDOUBLE(region.radius),
+        @"identifier" : region.identifier
+      };
+      [event setObject:dict forKey:@"region"];
+    }
+  }
 
   return event;
 }

--- a/iphone/Classes/TiApp.m
+++ b/iphone/Classes/TiApp.m
@@ -617,9 +617,9 @@ TI_INLINE void waitForMemoryPanicCleared(); //WARNING: This must never be run on
 {
   if ([[[[[response notification] request] content] userInfo] valueForKey:@"aps"] != nil) {
     NSMutableDictionary *responseInfo = nil;
-    if ([response isKindOfClass:[UNTextInputNotificationResponse class]]){
+    if ([response isKindOfClass:[UNTextInputNotificationResponse class]]) {
       responseInfo = [NSMutableDictionary dictionary];
-      [responseInfo setValue:((UNTextInputNotificationResponse*)response).userText forKey:UIUserNotificationActionResponseTypedTextKey];
+      [responseInfo setValue:((UNTextInputNotificationResponse *)response).userText forKey:UIUserNotificationActionResponseTypedTextKey];
     }
     [self application:[UIApplication sharedApplication] handleActionWithIdentifier:response.actionIdentifier forRemoteNotification:response.notification.request.content.userInfo withResponseInfo:responseInfo completionHandler:completionHandler];
   } else {
@@ -627,8 +627,8 @@ TI_INLINE void waitForMemoryPanicCleared(); //WARNING: This must never be run on
     RELEASE_TO_NIL(localNotification);
     localNotification = [[[self class] dictionaryWithUserNotification:response.notification
                                                        withIdentifier:response.actionIdentifier] retain];
-    if ([response isKindOfClass:[UNTextInputNotificationResponse class]]){
-       [localNotification setValue:((UNTextInputNotificationResponse*)response).userText forKey:@"typedText"];
+    if ([response isKindOfClass:[UNTextInputNotificationResponse class]]) {
+      [localNotification setValue:((UNTextInputNotificationResponse *)response).userText forKey:@"typedText"];
     }
     [self tryToPostNotification:localNotification withNotificationName:kTiLocalNotificationAction completionHandler:completionHandler];
   }
@@ -1490,25 +1490,25 @@ TI_INLINE void waitForMemoryPanicCleared(); //WARNING: This must never be run on
   if (category != nil) {
     event[@"category"] = category;
   }
-  NSArray* backgroundModes = [[NSBundle mainBundle] objectForInfoDictionaryKey:@"UIBackgroundModes"];
+  NSArray *backgroundModes = [[NSBundle mainBundle] objectForInfoDictionaryKey:@"UIBackgroundModes"];
   if ([backgroundModes containsObject:@"remote-notification"]) {
-      // Generate unique key with timestamp.
-      id key = [NSString stringWithFormat:@"CategoryPush-%f",[[NSDate date] timeIntervalSince1970]];
-      // Store the completionhandler till we can come back and send appropriate message.
-      if (pendingCompletionHandlers == nil) {
-          pendingCompletionHandlers = [[NSMutableDictionary alloc] init];
-      }
-      [pendingCompletionHandlers setObject:[[completionHandler copy] autorelease ]forKey:key];
+    // Generate unique key with timestamp.
+    id key = [NSString stringWithFormat:@"CategoryPush-%f", [[NSDate date] timeIntervalSince1970]];
+    // Store the completionhandler till we can come back and send appropriate message.
+    if (pendingCompletionHandlers == nil) {
+      pendingCompletionHandlers = [[NSMutableDictionary alloc] init];
+    }
+    [pendingCompletionHandlers setObject:[[completionHandler copy] autorelease] forKey:key];
 
-      NSMutableDictionary* dict = [NSMutableDictionary dictionaryWithObjectsAndKeys:key, @"handlerId", nil];
-      [dict addEntriesFromDictionary:event];
-      [self tryToPostBackgroundModeNotification:dict
-                           withNotificationName:kTiRemoteNotificationAction];
-      // We will go ahead and keeper a timer just in case the user returns the value too late - this is the worst case scenario.
-      NSTimer*  flushTimer = [NSTimer timerWithTimeInterval:TI_BACKGROUNDFETCH_MAX_INTERVAL target:self selector:@selector(fireCompletionHandler:) userInfo:key repeats:NO] ;
-      [[NSRunLoop mainRunLoop] addTimer:flushTimer forMode:NSDefaultRunLoopMode];
+    NSMutableDictionary *dict = [NSMutableDictionary dictionaryWithObjectsAndKeys:key, @"handlerId", nil];
+    [dict addEntriesFromDictionary:event];
+    [self tryToPostBackgroundModeNotification:dict
+                         withNotificationName:kTiRemoteNotificationAction];
+    // We will go ahead and keeper a timer just in case the user returns the value too late - this is the worst case scenario.
+    NSTimer *flushTimer = [NSTimer timerWithTimeInterval:TI_BACKGROUNDFETCH_MAX_INTERVAL target:self selector:@selector(fireCompletionHandler:) userInfo:key repeats:NO];
+    [[NSRunLoop mainRunLoop] addTimer:flushTimer forMode:NSDefaultRunLoopMode];
   } else {
-      [self tryToPostNotification:[event autorelease] withNotificationName:kTiRemoteNotificationAction completionHandler:completionHandler];
+    [self tryToPostNotification:[event autorelease] withNotificationName:kTiRemoteNotificationAction completionHandler:completionHandler];
   }
 }
 

--- a/iphone/Classes/TiApp.m
+++ b/iphone/Classes/TiApp.m
@@ -601,14 +601,14 @@ TI_INLINE void waitForMemoryPanicCleared(); //WARNING: This must never be run on
 // iOS 10+
 - (void)userNotificationCenter:(UNUserNotificationCenter *)center willPresentNotification:(UNNotification *)notification withCompletionHandler:(void (^)(UNNotificationPresentationOptions options))completionHandler
 {
-    // For backwards compatibility with iOS < 10, we do not show notifications in-app, but make it configurable
-    BOOL showInForeground = [TiUtils boolValue:notification.request.content.userInfo[@"showInForeground"] def:NO];
-    if (showInForeground) {
-      completionHandler(UNNotificationPresentationOptionBadge | UNNotificationPresentationOptionAlert | UNNotificationPresentationOptionSound);
-    } else {
-      [self application:[UIApplication sharedApplication] didReceiveRemoteNotification:notification.request.content.userInfo];
-      completionHandler(UNNotificationPresentationOptionNone);
-    }
+  // For backwards compatibility with iOS < 10, we do not show notifications in-app, but make it configurable
+  BOOL showInForeground = [TiUtils boolValue:notification.request.content.userInfo[@"showInForeground"] def:NO];
+  if (showInForeground) {
+    completionHandler(UNNotificationPresentationOptionBadge | UNNotificationPresentationOptionAlert | UNNotificationPresentationOptionSound);
+  } else {
+    [self application:[UIApplication sharedApplication] didReceiveRemoteNotification:notification.request.content.userInfo];
+    completionHandler(UNNotificationPresentationOptionNone);
+  }
 }
 
 // iOS 10+
@@ -623,7 +623,7 @@ TI_INLINE void waitForMemoryPanicCleared(); //WARNING: This must never be run on
     }
     [self application:[UIApplication sharedApplication] handleActionWithIdentifier:response.actionIdentifier forRemoteNotification:response.notification.request.content.userInfo withResponseInfo:responseInfo completionHandler:completionHandler];
   } else {
-      //NOTE Local notifications should be handled similar to BG above which ultimately calls handleRemoteNotificationWithIdentifier as this will allow BG Actions to execute.
+    //NOTE Local notifications should be handled similar to BG above which ultimately calls handleRemoteNotificationWithIdentifier as this will allow BG Actions to execute.
     RELEASE_TO_NIL(localNotification);
     localNotification = [[[self class] dictionaryWithUserNotification:response.notification
                                                        withIdentifier:response.actionIdentifier] retain];
@@ -1490,26 +1490,26 @@ TI_INLINE void waitForMemoryPanicCleared(); //WARNING: This must never be run on
   if (category != nil) {
     event[@"category"] = category;
   }
-    NSArray* backgroundModes = [[NSBundle mainBundle] objectForInfoDictionaryKey:@"UIBackgroundModes"];
-    if ([backgroundModes containsObject:@"remote-notification"]) {
-        // Generate unique key with timestamp.
-        id key = [NSString stringWithFormat:@"CategoryPush-%f",[[NSDate date] timeIntervalSince1970]];
-        // Store the completionhandler till we can come back and send appropriate message.
-        if (pendingCompletionHandlers == nil) {
-            pendingCompletionHandlers = [[NSMutableDictionary alloc] init];
-        }
-        [pendingCompletionHandlers setObject:[[completionHandler copy] autorelease ]forKey:key];
+  NSArray* backgroundModes = [[NSBundle mainBundle] objectForInfoDictionaryKey:@"UIBackgroundModes"];
+  if ([backgroundModes containsObject:@"remote-notification"]) {
+      // Generate unique key with timestamp.
+      id key = [NSString stringWithFormat:@"CategoryPush-%f",[[NSDate date] timeIntervalSince1970]];
+      // Store the completionhandler till we can come back and send appropriate message.
+      if (pendingCompletionHandlers == nil) {
+          pendingCompletionHandlers = [[NSMutableDictionary alloc] init];
+      }
+      [pendingCompletionHandlers setObject:[[completionHandler copy] autorelease ]forKey:key];
 
-        NSMutableDictionary* dict = [NSMutableDictionary dictionaryWithObjectsAndKeys:key, @"handlerId", nil];
-        [dict addEntriesFromDictionary:event];
-        [self tryToPostBackgroundModeNotification:dict
-                             withNotificationName:kTiRemoteNotificationAction];
-        // We will go ahead and keeper a timer just in case the user returns the value too late - this is the worst case scenario.
-        NSTimer*  flushTimer = [NSTimer timerWithTimeInterval:TI_BACKGROUNDFETCH_MAX_INTERVAL target:self selector:@selector(fireCompletionHandler:) userInfo:key repeats:NO] ;
-        [[NSRunLoop mainRunLoop] addTimer:flushTimer forMode:NSDefaultRunLoopMode];
-    } else {
-        [self tryToPostNotification:[event autorelease] withNotificationName:kTiRemoteNotificationAction completionHandler:completionHandler];
-    }
+      NSMutableDictionary* dict = [NSMutableDictionary dictionaryWithObjectsAndKeys:key, @"handlerId", nil];
+      [dict addEntriesFromDictionary:event];
+      [self tryToPostBackgroundModeNotification:dict
+                           withNotificationName:kTiRemoteNotificationAction];
+      // We will go ahead and keeper a timer just in case the user returns the value too late - this is the worst case scenario.
+      NSTimer*  flushTimer = [NSTimer timerWithTimeInterval:TI_BACKGROUNDFETCH_MAX_INTERVAL target:self selector:@selector(fireCompletionHandler:) userInfo:key repeats:NO] ;
+      [[NSRunLoop mainRunLoop] addTimer:flushTimer forMode:NSDefaultRunLoopMode];
+  } else {
+      [self tryToPostNotification:[event autorelease] withNotificationName:kTiRemoteNotificationAction completionHandler:completionHandler];
+  }
 }
 
 - (void)application:(UIApplication *)application

--- a/iphone/Classes/TiAppiOSUserNotificationCenterProxy.m
+++ b/iphone/Classes/TiAppiOSUserNotificationCenterProxy.m
@@ -226,11 +226,16 @@
   [event setObject:NULL_IF_NIL([[request content] title]) forKey:@"alertTitle"];
   [event setObject:NULL_IF_NIL([[request content] subtitle]) forKey:@"alertSubtitle"];
   [event setObject:NULL_IF_NIL([[request content] launchImageName]) forKey:@"alertLaunchImage"];
-  [event setObject:NULL_IF_NIL([[request content] sound]) forKey:@"sound"];
   [event setObject:NULL_IF_NIL([[request content] badge]) forKey:@"badge"];
   [event setObject:NULL_IF_NIL([[request content] userInfo]) forKey:@"userInfo"];
   [event setObject:NULL_IF_NIL([[request content] categoryIdentifier]) forKey:@"category"];
   [event setObject:NULL_IF_NIL([request identifier]) forKey:@"identifier"];
+
+  // iOS 10+ does have "soundName" but "sound" which is a native object. But if we find
+  // a sound in the APS dictionary, we can provide that one for parity
+  if (request.content.userInfo[@"aps"] && request.content.userInfo[@"aps"][@"sound"]) {
+    [event setObject:request.content.userInfo[@"aps"][@"sound"] forKey:@"sound"];
+  }
 
   if ([[request trigger] isKindOfClass:[UNCalendarNotificationTrigger class]]) {
     [event setObject:NULL_IF_NIL([(UNCalendarNotificationTrigger *)[request trigger] nextTriggerDate]) forKey:@"date"];

--- a/iphone/cli/commands/_build.js
+++ b/iphone/cli/commands/_build.js
@@ -199,7 +199,7 @@ iOSBuilder.prototype.assertIssue = function assertIssue(issues, name) {
 	for (let i = 0; i < issues.length; i++) {
 		if ((typeof name === 'string' && issues[i].id === name) || (typeof name === 'object' && name.test(issues[i].id))) {
 			this.logger.banner();
-			appc.string.wrap(issues[i].message, this.config.get('cli.width', 100)).split('\n').forEach(function (line, i, arr) {
+			appc.string.wrap(issues[i].message, this.config.get('cli.width', 100)).split('\n').forEach(function (line) {
 				this.logger.error(line.replace(/(__(.+?)__)/g, '$2'.bold));
 			}, this);
 			this.logger.log();

